Single storage of track angular fluxes for cases with vacuum boundary conditions only

GiudGiud opened the pull request:
This PR introduces the ONLYVACUUMBC macro, set at compilation time. 

Basically if the case only contains vacuum boundary conditions, it allows to store each track angular flux only once, thus saving a factor of two on memory on large cases, for which memory costs are clearly dominated by track angular flux storage.

Without domain decomposition, it's trivial
With domain decomposition, we have to be very careful during flux transfer time not to overwrite a flux that has not been sent yet. If an overwrite is about to happen, the flux is saved in the send_buffers to be transferred at the next round. This pre-filling of buffers can overflow them, when that happens, the buffer is doubled in size dynamically and a log is printed.

Communication costs overall go up, probably because the routine is now more complex and less optimized. 
Sweeping time remains dominant in the cases that are memory bound, for which we need something like this. 

Other changes :
- the code doesn't crash anymore when zero volume regions are used
- info logs now print the rank as well

GiudGiud commented:
@mit-crpg/openmoc-dev anyone to review this ? It's quite critical for 3D full core calculations on less than 100 nodes (at 128 gb RAM) to reduce the storage of boundary angular fluxes.

GiudGiud commented:
Merging with self-review Monday

wuwenbin2006 commented on src/CPUSolver.cpp:
```diff
@@ -490,27 +504,43 @@ void CPUSolver::setupMPIBuffers() {
+    MPI_Allreduce(&size_mb, &max_size_mb, 1, MPI_INT, MPI_MAX,
+                  _geometry->getMPICart());
+    log_printf(NORMAL, "Max track fluxes transfer buffer storage = %.2f MB",
+               max_size_mb / 1e6);
```
line 519~525 should be put after line 535. otherwise, size_mb is zero

GiudGiud replied:
Good catch. Not sure why it wasnt printing 0 for me

wuwenbin2006 commented on src/CPUSolver.cpp:
```diff
@@ -663,6 +717,35 @@ void CPUSolver::deleteMPIBuffers() {
+    }
+  }
+  else {
+    for (long t=0; t < _tot_num_tracks; t++)
```
OMP could be used here.

GiudGiud replied:
Agreed

wuwenbin2006 commented on src/CPUSolver.cpp:
```diff
@@ -821,8 +905,10 @@ void CPUSolver::packBuffers(std::vector<long> &packing_indexes) {
   for (int i=0; i < num_domains; i++) {
     int send_domain = _neighbor_domains.at(i);
```
Seems send_domain is used nowhere.

GiudGiud replied:
Good catch

wuwenbin2006 commented on src/CPUSolver.cpp:
```diff
@@ -821,8 +905,10 @@ void CPUSolver::packBuffers(std::vector<long> &packing_indexes) {
-    int start_idx = _fluxes_per_track + 1;
+    /* Reset send buffers : start at beginning if the buffer has not been
+       prefilled, else start after what has been prefilled */
+    int start_idx = (_send_buffers_index.at(i)) * (_track_message_size) +
```
two pairs of brackets are useless here. remove them.

GiudGiud replied:
Done

wuwenbin2006 commented on src/CPUSolver.cpp:
```diff
@@ -907,8 +1025,51 @@ void CPUSolver::transferAllInterfaceFluxes() {
+        reinterpret_cast<long*>(&_send_buffers.at(i)[_fluxes_per_track+1]);
+      long first_track = first_track_idx[0];
+      if (first_track != -1)
+        need_to_send = 1;
```
A break of the for loop could be put here avoiding extra computation.

GiudGiud replied:
There's not much work done in that loop and not many iterations, it's fine to keep going

GiudGiud commented:
@wuwenbin2006 thanks a lot for the review ! I think I addressed everything, and added an optimization for the track buffer size.  
In my current timings, angular flux exchange is 10% of the runtime for a full core case when using this special case of only-vacuum-boundary-conditions, so I would hold off on optimizing it for now. 

I agree the code is a bit clunky right now. We have to figure out a clean way to move some part of the solvers to another file. MPI communication routines are a good candidate. Once that's done, we can implement different ways of communicating across domains. Like I would like us to try a RMA implementation
Neighbor collective operations could be a good idea. Also, we have a cartesian mesh and if we find a MPI function that handles diagonal between domains as well we could get a pretty efficient implementation.

I think @friedmud 's PhD thesis will provide a lot of insight on how we can do angular flux exchange.
